ext/drreg/drreg.c

```diff
@@ -655,7 +655,6 @@ drreg_forward_analysis(void *drcontext, instr_t *start)
     for (reg = DR_REG_START_GPR; reg <= DR_REG_STOP_GPR; reg++) {
         pt->reg[GPR_IDX(reg)].app_uses = 0;
         drvector_set_entry(&pt->reg[GPR_IDX(reg)].live, 0, REG_UNKNOWN);
-        pt->reg[GPR_IDX(reg)].ever_spilled = false;
```

Comment:
> Resetting the spilled flag is not needed as it will be set in drreg_reserve_reg_internal().

If this is being relied upon, can we make it less fragile wrt future code changes.  Even now, it is not readily clear from the code: you have to figure out that already_spilled will never be set for all users of drreg_forward_analysis.  Could we add an assert or something that it's set?

Also, what about drreg_reserve_aflags() which also calls drreg_forward_analysis?

I'm looking at the history: you can see that this line was added in 8ee8d818e660756d1edb9e3be26198218a877386 and it must have been there for a reason.  Wondering if the test added there hit sthg w/o it.

Reply:
Agree. I was looking at this. I couldn't figure out why it was added. ever_spilled may be true due to earlier calls of drreg_reserve_register(). Each makes a call to drreg_forward_analysis(). So clearly resetting ever_spilled doesn't make sense, because then drreg_restore_app_value() will ignore. Which is what happened to me. When calling from insertion phase, ever_spilled get only reset during initialization. So as much as w/o forward analysis ever_spilled might be already set (already spilled), it may be set if not doing forward analysis either. I don't see how to make it more robust. But again disclaimer, I don't understand why it was added. Basically similar to what we discussed yesterday offline, I don't see the difference of called from insertion phase and called from outside insertion phase w.r.t. ever_spilled and why it would need to be reset every time for the latter.

Reply:
I'm trying to see where ever_spilled is reset for the insertion phase: shouldn't it be set to false at the start of a new block, in drreg_event_bb_analysis next to where app_uses=0 or sthg like that?  The insertion phase has nice control points for where to reset state: outside does not, which is likely why it was put here.  But, since this drreg_forward_analysis() is called even on query API calls, it should not be changing that state.

Where will ever_spilled be reset for the next bb, then?  Seems like drreg might want a very-early app2app event and reset it there, for both insertion and other phases.

Reply:
>> The insertion phase has nice control points for where to reset state
Do you see where it _is_ reset though for insertion phase? I don't.

Reply:
Or is a client supposed to call drreg_init every time?

Reply:
> Do you see where it is reset though for insertion phase? I don't.
That is exactly what my comment above is about: "I'm trying to see where ever_spilled is reset for the insertion phase".  Please re-read.

Reply:
Did read. I think it isn't, only reset in init (which is why I was asking https://github.com/DynamoRIO/dynamorio/pull/3822#discussion_r322870315). The docs mention that drreg_init() can be called multiple times, is that why?

Reply:
No, drreg_init initializes process-global state like the number of spill slots.  As the docs say, the reason for allowing multiple calls is for multiple libraries each using drreg.

Reply:
Then I would say this is a latent, additional bug.

Reply:
Yes, filed as #3827.  Unclear whether it has any visible consequences at all though.